File: bytedance/string/014. Longest Common Prefix.go
// Vertical
// Time complexity: O(mn) m is average length of strings. n is number of strings.
func longestCommonPrefix(strs []string) string {
    if len(strs) == 0 {
        return ""
    }

    for i := 0; i < len(strs[0]); i++ {
        for j := 0; j < len(strs); j++ {
            // Either strs[j] has the same length with strs[0] or the character are not equal.
            if i == len(strs[j]) || strs[j][i] != strs[0][i] { 
                return strs[0][:i]
            }
        }
    }

    return strs[0]
}

// One by one
// Time complexity: O(mn) m is average length of strings. n is number of strings.
func longestCommonPrefix(strs []string) string {
    if len(strs) == 0 {
        return ""
    }

    prefix := strs[0]

    for i := 0; i < len(strs); i++ {
        prefix = lcp(prefix, strs[i])
        if len(prefix) == 0 {
            break
        }
    }

    return prefix
}

func lcp(str1, str2 string) string {
    length := min(len(str1), len(str2))
    index := 0
    for index < length && str1[index] == str2[index] {
        index++
    }
    return str1[:index]
}

func min(n1, n2 int) int {
    if n1 < n2 {
        return n1
    }
    return n2
}

File: bytedance/string/014. Longest Common Prefix.py
# Solution 1 compare every character vertically in **two row**
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        r = [len(set(c)) == 1 for c in zip(*strs)] + [0]
        return str[0][:r.index(0)] if strs else ''


# Solution 2 compare every character vertically
class Solution:
    def longestCommonPrefix(self, strs):
        res = ''
        for i in zip(*str):
            tmp_set = set(i)
            if len(tmp_set) == 1:
                res += i[0]
            else:
                break
        return res

# Solution 3 compare one by one
class Solution:
    def longestCommonPrefix(self, s: List[str]) -> str:
        if not strs: return ''
        a = strs[0]
        for i in range(1, len(strs)):
            while strs[i].find(a) != 0:
                a = a[:len(a)-1]
        return a
# or
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not s: return ''

        prefix, count = s[0], len(s)
        for i in range(1, count):
            prefix = self.lcp(prefix, s[i])
            if not prefix:
                break
        return prefix

    def lcp(self, str1, str2):
        length, index = min(len(str1), len(str2)), 0
        while index < length and str1[index] == str2[index]:
            index += 1
        return str1[:index]

# Solution 4 compare the first and last one after sorting.
class Solution:
    def longestCommonPrefix(self, s: List[str]) -> str:
        if not s:
            return ""
        s.sort()
        a, b, res = s[0], s[-1], ''
        for i in range(len(a)):
            if a[i] == b[i]:
                res += a[i]
            else:
                break
        return res